### ISSUE TITLE
H.264 acceleration through VDPAU not working with "mpv --vo=gpu --hwdec=vdpau"

### ISSUE THREAD

**Issue by lorn10** [mpv-player/mpv#11120]
### Important Information

```
test@iMac-test:~/Videos$ mpv
mpv 0.34.1 Copyright © 2000-2021 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
FFmpeg library versions:
   libavutil       56.70.100
   libavcodec      58.134.100
   libavformat     58.76.100
   libswscale      5.9.100
   libavfilter     7.110.100
   libswresample   3.9.100
FFmpeg version: 4.4.2-0ubuntu0.22.04.1

```
```
test@test-iMac:~$ inxi -GSC -xx
System:
  Host: test-iMac Kernel: 5.15.0-57-generic x86_64 bits: 64 compiler: gcc
    v: 11.3.0 Desktop: KDE Plasma 5.24.7 tk: Qt 5.15.3 wm: kwin_x11 dm: SDDM
    Distro: Ubuntu 22.04.1 LTS (Jammy Jellyfish)
CPU:
  Info: quad core model: Intel Core i5-2400 bits: 64 type: MCP
    arch: Sandy Bridge rev: 7 cache: L1: 256 KiB L2: 1024 KiB L3: 6 MiB
  Speed (MHz): avg: 1600 min/max: 1600/3400 cores: 1: 1600 2: 1600 3: 1600
    4: 1600 bogomips: 24802
  Flags: avx ht lm nx pae sse sse2 sse3 sse4_1 sse4_2 ssse3 vmx
Graphics:
  Device-1: Intel 2nd Generation Core Processor Family Integrated Graphics
    vendor: Apple driver: i915 v: kernel ports: active: none empty: VGA-1
    bus-ID: 00:02.0 chip-ID: 8086:0102
  Device-2: AMD Whistler [Radeon HD 6730M/6770M/7690M XT] vendor: Apple
    driver: radeon v: kernel pcie: speed: 2.5 GT/s lanes: 16 ports:
    active: eDP-1 empty: DP-1, DP-2, DP-3, DP-4, VGA-2 bus-ID: 01:00.0
    chip-ID: 1002:6740
  Device-3: Apple FaceTime HD Camera (Built-in) type: USB driver: uvcvideo
    bus-ID: 1-2:3 chip-ID: 05ac:850b
  Display: x11 server: X.Org v: 1.21.1.3 compositor: kwin_x11 driver: X:
    loaded: ati,modesetting,radeon unloaded: fbdev,vesa gpu: radeon
    display-ID: :0 screens: 1
  Screen-1: 0 s-res: 2560x1440 s-dpi: 96
  Monitor-1: eDP res: 2560x1440 dpi: 109 diag: 685mm (27")
  OpenGL: renderer: AMD TURKS (DRM 2.50.0 / 5.15.0-57-generic LLVM 15.0.5)
    v: 4.5 Mesa 23.0.0-devel (git-8452709 2023-01-09 jammy-oibaf-ppa)
```

### Reproduction steps

When a H.264 file is played in mpv and VDPAU acceleration enabled it falls always back to software decoding. 

Please use the **Big Buck Bunny H.264** test file for testing (see download link below).

### Expected behavior

`mpv --vo=gpu --hwdec=vdpau` should work.

### Actual behavior

`mpv --vo=gpu --hwdec=vdpau` doesn't work. Instead of it `mpv --vo=vdpau --hwdec=vdpau` has to be used which is itself not recommended by mpv:

```
test@iMac-test:~/Videos$ mpv --vo=vdpau --hwdec=vdpau big_buck_bunny_1080p_h264_AAC.mov
 (+) Video --vid=1 (*) (h264 1920x1080 24.000fps)
 (+) Audio --aid=1 --alang=eng (*) (aac 6ch 48000Hz)
[vo/vdpau] Warning: this compatibility VO is low quality and may have issues with OSD, scaling, screenshots and more.
[vo/vdpau] vo=gpu is the preferred choice in any case and includes VDPAU support via hwdec=vdpau or vdpau-copy.
Using hardware decoding (vdpau).
AO: [pulse] 48000Hz 5.1 6ch float
VO: [vdpau] 1920x1080 vdpau[yuv420p]
[vo/vdpau] Compositing window manager detected. Assuming timing info is inaccurate.
AV: 00:00:05 / 00:09:56 (1%) A-V:  0.000

Exiting... (Quit)
```
There is also an error message `Cannot load libcuda.so.1` displayed at the CLI when the `--vo=gpu` setting is used. This doesn't happens when the `--vo=vdpau` parameter is set. For whatever reason mpv is looking for that file even at a **non-NVIDIA** based systems. Maybe this triggers the fallback to software decoding.

### Log file
```
[vo/gpu/opengl] Loaded extension GL_NV_vdpau_interop.
[vo/gpu] Testing FBO format rgba16f
[vo/gpu] Using FBO format rgba16f.
[vo/gpu] No advanced processing required. Enabling dumb mode.
[vo/gpu] Assuming 59.950550 FPS for display sync.
[vd] Container reported FPS: 24.000000
[vd] Codec list:
[vd]     h264 - H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
[vd]     h264_v4l2m2m (h264) - V4L2 mem2mem H.264 decoder wrapper
[vd]     h264_qsv (h264) - H264 video (Intel Quick Sync Video acceleration)
[vd]     h264_cuvid (h264) - Nvidia CUVID H264 decoder
[vd] Opening decoder h264
[vd] Looking at hwdec h264-vdpau...
[vo/gpu] Loading hwdec driver 'vaapi-egl'
[vo/gpu/vaapi-egl] using VAAPI EGL interop
[vo/gpu/vaapi-egl] Trying to open a x11 VA display...
[vo/gpu/vaapi-egl/vaapi] Initialized VAAPI: version 1.14
[vo/gpu/vaapi-egl] Going to probe surface formats (may log bogus errors)...
[vo/gpu/vaapi-egl] Done probing surface formats.
[vo/gpu] Loading hwdec driver 'cuda-nvdec'
Cannot load libcuda.so.1
[vo/gpu/cuda-nvdec] Failed to load CUDA symbols
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'drmprime-drm'
[vo/gpu/drmprime-drm] Failed to retrieve DRM fd from native display.
[vo/gpu] Loading failed.
[vd] Could not create device.
[vd] No hardware decoding available for this codec.
[vd] Using software decoding.
```

Full log file: 
[mpv_log_gpu_vdpau_H264.txt](https://github.com/mpv-player/mpv/files/10381318/mpv_log_gpu_vdpau_H264.txt)

### Sample files

You can use for testing the original **Big Buck Bunny H.264** video file from here:
https://www.chip.de/downloads/Big-Buck-Bunny-Gratis-Kurzfilm_31870850.html

**Comment by lorn10** [comment]
As a last example the situation is similar with the `--vo=gpu --hwdec=vaapi` paramters. Also here mpv prefers to use as first the non-existent `cuda-nvdec` decoder. But in contrast to the previous codecs the mpv fallback is this time better handled. 

It falls firsst back to `drmprime-drm` which also fails and after that it selects `h264-vaapi` which then finally works.

```
[vo/gpu] Loading hwdec driver 'vaapi-egl'
[vo/gpu/vaapi-egl] using VAAPI EGL interop
[vo/gpu/vaapi-egl] Trying to open a x11 VA display...
[vo/gpu/vaapi-egl/vaapi] Initialized VAAPI: version 1.14
[vo/gpu/vaapi-egl] Going to probe surface formats (may log bogus errors)...
[vo/gpu/vaapi-egl] Done probing surface formats.
[vo/gpu] Loading hwdec driver 'cuda-nvdec'
Cannot load libcuda.so.1
[vo/gpu/cuda-nvdec] Failed to load CUDA symbols
[vo/gpu] Loading failed.
[vo/gpu] Loading hwdec driver 'drmprime-drm'
[vo/gpu/drmprime-drm] Failed to retrieve DRM fd from native display.
[vo/gpu] Loading failed.
[vd] Trying hardware decoding via h264-vaapi.
[vd] Selected codec: h264 (H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10)
[vf] User filter list:
[vf]   (empty)
[ad] Codec list:
[ad]     aac - AAC (Advanced Audio Coding)
[ad]     aac_fixed (aac) - AAC (Advanced Audio Coding)
[ad] Opening decoder aac
[ad] Requesting 1 threads for decoding.
[ad] Selected codec: aac (AAC (Advanced Audio Coding))
[af] User filter list:
[af]   (empty)
[cplayer] Starting playback...
[cplayer] Set property: shared-script-properties -> 1
[vd] Pixel formats supported by decoder: vdpau cuda vaapi_vld yuv420p
[vd] Codec profile: Main (0x4d)
[vd] Requesting pixfmt 'vaapi_vld' from decoder.
[vd] Using hardware decoding (vaapi).
```

So in the end this topic could be probably easy solved by **changing the mpv hardware decoder preference** under Linux to something like:

1. **VA-API**
2. **VDPAU**
3. **cuda-nvdec**
4. **drmprime-drm** (or whatever is meant here)
5. **pure software decoding**

**Comment by Dudemanguy** [comment]
First of all, please update your mpv to 0.35 if possible. There were quite a lot of hardware decoding changes since then and I believe that solves the autoprobing issues. For `--hwdec=vdpau`, I think you have to explicitly use glx instead of egl which isn't the default (`--gpu-context=x11`) and actually this disabled in the build so this option might not be available to you depending on the distro.

**Comment by Dudemanguy** [comment]
Yeah the unnecessary `cuda-nvdec` probing is resolved in mpv 0.35.

Just a couple of other things. `vdpau` is pretty much dead at this point and behind other hwdec solutions. mpv still supports it for legacy reasons mainly, but in general users shouldn't use this. If you are on intel/amd, you should be using `vaapi`. NVIDIA nowadays prefers `nvdec` (which mpv supports). `vdpau` is pretty much only useful for legacy nvidia cards.

**Comment by lorn10** [comment]
Great :+1: , and here follows for the record the relevant output of the working `mpv --vo=vdpau --hwdec=vdpau` setting:

```
[vo/vdpau/x11] X11 opening display: :0
[vo/vdpau/x11] Display 0 (eDP): [0, 0, 2560, 1440] @ 59.950550 FPS
[vo/vdpau/x11] Current display FPS: 59.950550
[vo/vdpau] Warning: this compatibility VO is low quality and may have issues with OSD, scaling, screenshots and more.
[vo/vdpau] vo=gpu is the preferred choice in any case and includes VDPAU support via hwdec=vdpau or vdpau-copy.
[vo/vdpau] Assuming 59.950550 FPS for display sync.
[vd] Container reported FPS: 24.000000
[vd] Codec list:
[vd]     h264 - H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
[vd]     h264_v4l2m2m (h264) - V4L2 mem2mem H.264 decoder wrapper
[vd]     h264_qsv (h264) - H264 video (Intel Quick Sync Video acceleration)
[vd]     h264_cuvid (h264) - Nvidia CUVID H264 decoder
[vd] Opening decoder h264
[vd] Looking at hwdec h264-vdpau...
[vd] Trying hardware decoding via h264-vdpau.
[vd] Selected codec: h264 (H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10)
[vf] User filter list:
[vf]   (empty)
[ad] Codec list:
[ad]     aac - AAC (Advanced Audio Coding)
[ad]     aac_fixed (aac) - AAC (Advanced Audio Coding)
[ad] Opening decoder aac
[ad] Requesting 1 threads for decoding.
[ad] Selected codec: aac (AAC (Advanced Audio Coding))
[af] User filter list:
[af]   (empty)
[cplayer] Starting playback...
[vd] Pixel formats supported by decoder: vdpau cuda vaapi_vld yuv420p
[vd] Codec profile: Main (0x4d)
[vd] Requesting pixfmt 'vdpau' from decoder.
[vd] Using hardware decoding (vdpau).
```

**PS** Yes, VDPAU is most likely deprecated in the foreseeable future but it makes definitely sense to have it especially for older Mesa drivers like [r300](https://gitlab.freedesktop.org/mesa/mesa/-/issues/?label_name%5B%5D=r300) or nv30. And some "big endian" retro computing folks even use mpv on super ancient G4 and G5 Apple hardware. Whatever, but the most amazing thing is that the long long broken MPEG-2 hardware acceleration in Mesa might be finally fixed. The PowerPC G4 and G5 retro users are eager to have it. :wink:
